java/algorithm/src/main/java/org/apache/arrow/algorithm/deduplicate/VectorDeduplicator.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.deduplicate;
+
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.util.DataSizeRoundingUtil;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.BitVectorHelper;
+import org.apache.arrow.vector.IntVector;
+import org.apache.arrow.vector.ValueVector;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Remove adjacent equal elements from a vector.
+ * If the vector is sorted, it removes all duplicated values in the vector.
+ * @param <V> vector type.
+ */
+public class VectorDeduplicator<V extends ValueVector> implements AutoCloseable {
```

Comment:
@liyafan82 this is useful functionality, but I'm not sure about the interface.  At least for me the functionality seems like it is intended for some form of RLE. It seems like it might pay to make a utility class with static methods of something like (I name the methods assuming RLE):
```
static void populateRunStartIndicators(ValueVector inputVector, BitVector runStarts);
void populateRunLengths(BitVector runStarts,  IntVector runLengths );
// This might belong in a separate class, or there might be some existing funtionality in the code base
void selectValues(BitVector indicators, ValueVector inputVector, ValueVector outputVector);
```
At least to me its not clear from the documentation how this class class is supposed to be used.

Reply:
@emkornfield 
I try to make this API general, and RLE is an important application of it.
In addition, deduplication is also intensively used in SQL runtime.

I implemented it as an class object, instead of using static methods, because the bit set to hold start positions is an internal state, so IMO, it should not be exposed to the outside.

However, I have also provided a utility class with utility methods as you suggested, to support more flexible scenarios.